### PR TITLE
Added a cached loader for xload

### PR DESCRIPTION
**Pull request by sonnes**

### CODE REVIEW

File: xload/providers/cached/cache.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	Set(key, value string, ttl time.Duration) error
+}
+
+// MapCache is a simple cache implementation using a map.
+type MapCache struct {
+	m   map[string]string
+	ttl map[string]time.Time
+
+	now func() time.Time
+	mu  sync.Mutex
```
Comment (ajatprabha) [inline xload/providers/cached/cache.go:22]: Why not use a `sync.Map` with a wrapper struct?
Equivalent to `map[string]*mv`.

```go
type mv struct {
	val string
	ttl time.Time
}
```

File: xload/providers/cached/loader.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+		if v != "" {
+			return v, nil
+		}
+
+		loaded, err := l.Load(ctx, key)
+		if err != nil {
+			return "", err
+		}
+
+		// DESIGN: If the loader returns an empty value, we
```
Comment (ajatprabha) [inline xload/providers/cached/loader.go:30]: Can we make this configurable via an `Option` say `DisableEmptyValueHit`. Default can remain the same as now.

```go
var DisableEmptyValueHit = emptyValHit(false)

type emptyValHit bool
func (e emptyValHit) apply(o *options) { o.emptyHit = bool(e) }
```

**Review by ajatprabha — CHANGES_REQUESTED** [review]

File: xload/providers/cached/cache.go (hunk 1 above)
Comment (sonnes) [reply xload/providers/cached/cache.go]: Makes sense. Changed it

File: xload/providers/cached/loader.go (hunk 2 above)
Comment (sonnes) [reply xload/providers/cached/loader.go]: Added a functional option

File: xload/providers/cached/options.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+func (t TTL) apply(o *options) { o.ttl = time.Duration(t) }
+
+// Cache sets the cache implementation for the loader.
+func Cache(c Cacher) Option {
+	return optionFunc(func(o *options) { o.cache = c })
+}
+
+// DisableEmptyValueHit disables caching of empty values.
+func DisableEmptyValueHit() Option {
```
Comment (ajatprabha) [inline xload/providers/cached/options.go:25]: nit: Do we need it to be a function call? Can change it to a global variable also, that also adds type info. As I suggested in previous review.

**Review by ajatprabha — APPROVED** [review]

File: xload/providers/cached/options.go (hunk 3 above)
Comment (ajatprabha) [reply xload/providers/cached/options.go:25]:
```go
// DisableEmptyValueHit disables caching of empty values.
var DisableEmptyValueHit Option = emptyValueHit(false)

type emptyValueHit bool

func (e emptyValueHit) apply(o *options) { o.emptyValueHit = bool(e) }
```

```diff
--- cachedLoader := NewLoader(loader, TTL(ttl), Cache(mc), DisableEmptyValueHit())
+++ cachedLoader := NewLoader(loader, TTL(ttl), Cache(mc), DisableEmptyValueHit)
```

**Pull request merged**